proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,135 @@
+  - name: FileSimilarityRetriever
+    type: FileSimilarityRetriever
+    params:
+      document_store: DocumentStore
+      primary_retriever: EmbeddingRetriever
+      top_k: 30
+      file_aggregation_key: file_id
+      max_num_queries: 50
+
+pipelines:
+  - name: query
+    nodes:
+      - name: FileSimilarityRetriever
+        inputs: [Query]
+
+  - name: indexing
+    nodes:
+      - name: TextConverter
+        inputs: [File]
+
+      - name: EmbeddingRetriever
+        inputs: [TextConverter]
+
+      - name: DocumentStore
+        inputs: [EmbeddingRetriever]
+
+```
+
+# Motivation
+
+Initially this retriever was implemented by Mathis for a customer case to quickly retrieve similar files given a file as input. The reason we would like to add it to Haystack is because we see wider use for this node in future customer and community cases.
+
+One example where this could be useful is in academic writing, if you have a large number of scientific sources (journals, books, conference proceedings) stored and you'd like to find all that are similar to a specific article or to what you've written so far (in the latter case, it can also be helpful for plagiarism detection). And as for industry use cases, report writing in any field can be facilitated by file similarity retrieval. It can also be very helpful for lawyers preparing a case as well as journalists doing research for their articles.
+
+# Detailed design
+
+You can see the code for this proposal here: https://github.com/deepset-ai/haystack/pull/5666/files.
+
+As a general description, the FileSimilarityRetriever works by:
+1. Getting all documents corresponding to the provided file ID from the document store. The documents can be obtained together with their embeddings, to save resources by avoiding re-calculating them at query time.
+2. Retrieving similar docs to each document from the file, using one or two chosen retrieval methods.
+3.  a) Returning the top-k retrieved similar documents, if only one retrieval method was used.
+    b) Returning an aggregated list of similar documents retrieved by both retrievers, if two retrieval methods were chosen. The results are aggregated based on the reciprocal rank fusion score, though it is also possible to keep the original score (if one of the retrievers assigned the document a relevance score) in the metadata.
+    Reciprocal rank fusion is calculated as 1 / (k + index)
+    k is set to 61 in reciprocal rank fusion based on a University of Waterloo paper (co-authored with Google) called "Reciprocal Rank Fusion outperforms Condorcet and individual Rank Learning Methods" [https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf] where k=60 was suggested, and 1 was added as python lists are 0-based and the paper used 1-based ranking.
+4. Returning the top-k similar files, based on an aggregated relevance score of all documents pertaining to the file. In the results the user sees the most relevant document from each file.
```

Comment:
It isn't clear to me how the aggregation works. As far as I can see, aggregation occurs across two (three) axes:
1. Aggregation across documents from query file: I don't really understand how this is done. Do we just calculate the max or sum them? If we were to sum them, we could turn this from O(m*n) to O(m+n) due to distributivity of vector addition and dot product. (m is the number of documents per query and n the number of documents across which we search).
2. Aggregation across documents from same file found by similarity: In another comment, it's mentioned that this is max. I can see that this is probably the easiest approach because there is probably a different number of documents per file. However, it's also easy to imagine a case where this might fail. Take, for example, the research paper search: The related works section might mention an approach that is very similar to the query. However, you would be more interested in a paper where most documents are somewhat similar to your query because that would be the actual paper that is referred to in the related works section of the less related paper which would be returned just because the related works section is more concise and therefore only one document.
3. (There is also an aggregation step across multiple retrievers, but that seems clear to me.)

I am not saying that it's necessary to evaluate all possible aggregation approaches, but the proposal could perhaps explain which approaches are taken and why they are taken.

Reply:
It's largely based on this: https://arxiv.org/pdf/2201.01614.pdf

With the thought that the benefit of VRRF doesn't justify the complexity.

Reply:
Aggregation is done on both axes via reciprocal rank fusion.

Reply:
I think first we find the most relevant docs for each doc of the given file (with 1 or 2 retrievers), then calculate RRF for all retrieved documents, then sum up the RRFs of all documents pertaining to the same file to sort the final files by relevance, and return the top scoring document from each file in the final results. Seems the usage of RRF should mitigate cases where only one document of the file is highly relevant while the rest are not, since from that file only one document will be retrieved, while for a file that's consistently similar to query file across many of its documents, many documents will make it to the results and the score will go further up. 

But I'm not sure - @mathislucka could you please comment on the score aggregation method & why it was chosen?

Reply:
Thanks, that makes a lot of sense. Perhaps, it would be useful to link the paper in the proposal.

Reply:
Good idea! Added 👍